src/java.base/share/classes/java/text/ListFormat.java

```diff
@@ -0,0 +1,594 @@
+     * @param obj the object to check, {@code null} returns {@code false}
+     * @return {@code true} if this is equals to the other {@code ListFormat}
+     */
+    @Override
+    public boolean equals(Object obj) {
+        if (this == obj) {
+            return true;
+        }
+
+        if (obj instanceof ListFormat other) {
+            return locale.equals(other.locale) &&
+                Arrays.equals(patterns, other.patterns);
+        }
+
+        return false;
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public int hashCode() {
+        return Objects.hash(locale, Arrays.hashCode(patterns));
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public String toString() {
+        return
+            """
+            ListFormat [locale: "%s", start: "%s", middle: "%s",  end: "%s", two: "%s", three: "%s"]
+            """.formatted(locale.getDisplayName(), patterns[START], patterns[MIDDLE], patterns[END], patterns[TWO], patterns[THREE]);
+    }
+
+    private MessageFormat generateMessageFormat(Object[] input) {
+        var len = input.length;
+        return switch (len) {
+            case 0 -> throw new IllegalArgumentException("There should at least be one input string");
+            case 1 -> new MessageFormat("{0}", locale);
+            case 2, 3 -> new MessageFormat(patterns[len + 1], locale);
+            default -> new MessageFormat(createMessageFormatString(len), locale);
+        };
+    }
+
+    private String createMessageFormatString(int count) {
+        var sb = new StringBuilder(256).append(patterns[START]);
+        IntStream.range(2, count - 1).forEach(i -> sb.append(middleBetween).append("{").append(i).append("}"));
+        sb.append(patterns[END].replaceFirst("\\{0}", "").replaceFirst("\\{1}", "\\{" + (count - 1) + "\\}"));
+        return sb.toString();
+    }
+
+    @java.io.Serial
+    private void readObject(ObjectInputStream stream)
+            throws IOException, ClassNotFoundException {
+        stream.defaultReadObject();
+        try {
+            init();
+        } catch (IllegalArgumentException iae) {
+            throw new IOException("Deserialization failed.", iae);
+        }
+    }
+
+    /**
+     * A ListFormat type - {@link #STANDARD STANDARD}, {@link #OR OR}, and
+     * {@link #UNIT UNIT}.
+     * <p>
+     * {@code Type} is an enum which represents the type for formatting
+     * a list within a given {@code ListFormat} instance.
+     */
+    public enum Type {
+
+        /**
+         * The {@code STANDARD} ListFormat style. This is the default
+         * type, which concatenates elements in "and" enumeration.
+         */
+        STANDARD,
+
+        /**
+         * The {@code OR} ListFormat style. This style concatenates
+         * elements in "or" enumeration.
+         */
+        OR,
+
+        /**
+         * The {@code UNIT} ListFormat style. This style concatenates
+         * elements, useful for enumerating units.
+         */
```

Comment:
The word "style" used in Type, I assume you meant "type"? Just that it might be confused with Style below.

Same as previous comments, a combination of Type and Style, if I understand correctly, forms a specific pattern. I might say something about it in the enum class description.

A STANDARD type then is a simple list with conjunction "and", or an inclusive list, and etc.

Reply:
Ah, good catch. Will correct those style/type typos.

> A STANDARD type then is a simple list with conjunction "and", or an inclusive list, and etc.

I could not quite catch what you meant by this. Can you please elaborate on it more?

Reply:
In a List Pattern, Type, as in the current document, defined its type as in the language, while Style the writing style. I guess I would try to avoid stating it "concatenates elements", but rather represents a simple list with conjunction "and", that is also called an inclusive list as in the language term.